src/compiler/program.ts

```diff
@@ -211,6 +221,77 @@ namespace ts {
             return classifiableNames;
         }
 
+        function tryReuseStructureFromOldProgram(): boolean {
+            if (!host.hasChanges) {
+                // host does not support method 'hasChanges'
+                return false;
+            }
+            if (!oldProgram) {
+                return false;
+            }
+            
+            Debug.assert(!oldProgram.structureIsReused);
+
+            // there is an old program, check if we can reuse its structure
+            let oldRootNames = oldProgram.getRootFileNames();
+            if (rootNames.length !== oldRootNames.length) {
+                // different amount of root names - structure cannot be reused
+                return false;
+            }
+            
+            for (let i = 0; i < rootNames.length; i++) {
+                if (oldRootNames[i] !== rootNames[i]) {
+                    // different order of root names - structure cannot be reused
+                    return false;
+                }
+            }
+            
+            // check if program source files has changed in the way that can affect structure of the program
+            let newSourceFiles: SourceFile[] = [];
+            for (let oldSourceFile of oldProgram.getSourceFiles()) {
+                let newSourceFile: SourceFile;
+                if (host.hasChanges(oldSourceFile)) {

```

Comment:
how about doing referential comparison to know if anything changed instead of asking the host? e.g.

```ts
newSourceFile = host.getSourceFile(oldSourceFile.fileName, options.target);
if (newSourceFile !== oldSourceFile) {
    if (!newSourceFile) {
         return false;
    }
     // check tripleslash references
    ....
     // check imports
     .....
}
else {
    // file has no changes - use it as is
    newSourceFile = oldSourceFile;
}
```